libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,354 @@
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/ast/ASTUtils.h>
+#include <libsolidity/ast/ASTVisitor.h>
+#include <libsolidity/interface/ReadFile.h>
+#include <libsolidity/interface/StandardCompiler.h>
+#include <libsolidity/lsp/LanguageServer.h>
+
+#include <liblangutil/SourceReferenceExtractor.h>
+#include <liblangutil/CharStream.h>
+
+#include <libsolutil/Visitor.h>
+#include <libsolutil/JSON.h>
+
+#include <boost/filesystem.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+
+#include <fmt/format.h>
+
+#include <ostream>
+#include <string>
+
+using namespace std;
+using namespace std::placeholders;
+
+using namespace solidity::lsp;
+using namespace solidity::langutil;
+using namespace solidity::frontend;
+
+namespace
+{
+
+Json::Value toJson(LineColumn _pos)
+{
+	Json::Value json = Json::objectValue;
+	json["line"] = max(_pos.line, 0);
+	json["character"] = max(_pos.column, 0);
+
+	return json;
+}
+
+Json::Value toJsonRange(LineColumn const& _start, LineColumn const& _end)
+{
+	Json::Value json;
+	json["start"] = toJson(_start);
+	json["end"] = toJson(_end);
+	return json;
+}
+
+optional<LineColumn> parseLineColumn(Json::Value const& _lineColumn)
+{
+	if (_lineColumn.isObject() && _lineColumn["line"].isInt() && _lineColumn["character"].isInt())
+		return LineColumn{_lineColumn["line"].asInt(), _lineColumn["character"].asInt()};
+	else
+		return nullopt;
+}
+
+constexpr int toDiagnosticSeverity(Error::Type _errorType)
+{
+	// 1=Error, 2=Warning, 3=Info, 4=Hint
+	switch (Error::errorSeverity(_errorType))
+	{
+	case Error::Severity::Error: return 1;
+	case Error::Severity::Warning: return 2;
+	case Error::Severity::Info: return 3;
+	}
+	return 1;
+}
+
+}
+
+LanguageServer::LanguageServer(Transport& _transport):
+	m_client{_transport},
+	m_handlers{
+		{"$/cancelRequest", [](auto, auto) {/*nothing for now as we are synchronous */} },
+		{"cancelRequest", [](auto, auto) {/*nothing for now as we are synchronous */} },
+		{"exit", [this](auto, auto) { m_exitRequested = true; }},
+		{"initialize", bind(&LanguageServer::handleInitialize, this, _1, _2)},
+		{"initialized", [](auto, auto) {} },
+		{"shutdown", [this](auto, auto) { m_shutdownRequested = true; }},
+		{"textDocument/didChange", bind(&LanguageServer::handleTextDocumentDidChange, this, _1, _2)},
+		{"textDocument/didClose", [](auto, auto) {/*nothing for now*/}},
```

Comment:
It is actually expected from the server to publish diagnostics on all files in the project, no matter if they are open or not. "open" in this context just means that he client "owns" the state of the file, i.e. if it changes on disk, these changes should be ignored.

If the IDE supports projects, it can run a new LSP for each open project, by the way.

Reply:
> It is actually expected from the server to publish diagnostics on all files in the project, no matter if they are open or not.

You mean publish everything on every request of just be able to answer about any file in the project when the client asks for it?

Reply:
The server owns the diagnostics and it's the server's decision when to publish it.

Reply:
But we do need some way to purge files that are no longer present.

Reply:
That is true. Do we need to watch the filesystem in that case?

Reply:
I think [`DidChangeWatchedFiles` Notification](https://microsoft.github.io/language-server-protocol/specification#workspace_didChangeWatchedFiles) is what we need here.

> Servers are allowed to run their own file system watching mechanism and not rely on clients to provide file system events. However this is not recommended due to the following reasons: